### ISSUE TITLE
{{on}} modifier should not add event listeners if the passed function is undefined

### ISSUE THREAD

**Issue by NullVoxPopuli** [emberjs/rfcs#660]
the On modifier's addEventListener helper function: https://github.com/emberjs/ember.js/blob/a493c6a840ae6e56a612237c2afa495ed2cdfd50/packages/%40ember/-internals/glimmer/lib/modifiers/on.ts#L211

_always_ adds an event listener (here: https://github.com/emberjs/ember.js/blob/a493c6a840ae6e56a612237c2afa495ed2cdfd50/packages/%40ember/-internals/glimmer/lib/modifiers/on.ts#L392 )

I think it'd help out _a lot_ of folks to make this change to `update`:

https://github.com/emberjs/ember.js/blob/a493c6a840ae6e56a612237c2afa495ed2cdfd50/packages/%40ember/-internals/glimmer/lib/modifiers/on.ts#L374

```diff
    // use prior state values for removal
    removeEventListener(element, eventName, callback, options);

    // read updated values from the state object
-   addEventListener(state.element, state.eventName, state.callback, state.options);
+   if (state.callback) {
+     addEventListener(state.element, state.eventName, state.callback, state.options);
+    }
```

**Comment by buschtoens** [comment]
I already suggested that in the past and the recommendation was to conditionally apply the modifier once the RFC that allows passing around modifiers lands. 

I don't have the links at hand right now, because I'm on mobile.

**Comment by NullVoxPopuli** [comment]
I think that could def help, but I believe it to be mostly orthogonal to this issue?

**Comment by buschtoens** [comment]
I agree, but IIRC the line of reasoning was that not calling the modifier in the first place is more efficient then calling it and have it effectively no-op.

Wit the other RFC in place, you could basically use an `{{if}}` expression in place of the modifier:

```hbs
<button {{if @handler (modifier "on" "click" @handler)}}>
```

**Comment by jelhan** [comment]
Wondering if this should be considered a bug? Attaching an event listener with a listener, which is neither a function nor implements `EventListener` interface is a bug. The `{{on}}` modifier should prevent this.

But it's another question if it should support `undefinded` or any other value, which isn't a valid listener. Maybe it should just fail early with a helpful error message? To be honest I'm tending towards that solution.

Silently not applying an event listener is error prune and hard to debug. I can imagine myself spending hours until I notice that there is a typo. 🙈

But I agree that core misses a simple pattern to conditionally apply an event listener. An alternative solution might be returning a noop function if `{{fn}}` helper is called without any argument. Currently an error is thrown in that case (with an error message which isn't very helpful by the way).

I'm not sure about the performance impact of applying unneeded event listeners with a noop function. But I expect it to be acceptable for most cases. Also it would be more flexible as that pattern could be used at all places which expects a function - and not only for `{{on}}` modifier.

**Comment by sukima** [comment]
This isn't a good option but I do this:

```hbs
{{on "click" (if @foo @foo this.noop)}}
```

I don't know seems silly to bike shed when there is no solution landed anyway. I really dislike dropping ideas in favor of future work which is not available yet. It is like saying "that thing you needed is not important enough for a solution today." 🤷

**Comment by lukemelia** [comment]
Another workaround, which we use, is the [optional helper from ember-composable-helpers](https://github.com/DockYard/ember-composable-helpers#optional):

```hbs
{{on "click" (optional @foo)}}
```

**Comment by buschtoens** [comment]
Here's the RFC I was referring to: #432 [Contextual Helpers and Modifiers (a.k.a. "first-class helpers/modifiers") ](https://emberjs.github.io/rfcs/0432-contextual-helpers.html)
And here's the tracking issue: https://github.com/emberjs/rfc-tracking/issues/6

**Comment by buschtoens** [comment]
We also use `(optional)`, like @lukemelia suggested (https://github.com/emberjs/rfcs/issues/660#issuecomment-689301031) and I think it is superior to making the callback parameter itself optional, like suggested by @NullVoxPopuli in the OP https://github.com/emberjs/rfcs/issues/660#issue-695967735.

As @jelhan has already pointed out (https://github.com/emberjs/rfcs/issues/660#issuecomment-689014252), this would masquerade typos.

> An alternative solution might be returning a noop function if `{{fn}}` helper is called without any argument.

This would pose the exact same hazard, just in a different place.

IMO using `(optional)` is the perfect solution, while RFC #432 is still pending implementation. It's a bit unfortunate that it's not part of Ember Core, but I think that's okay.

**Comment by NullVoxPopuli** [comment]
Atm, there are two related goals
- clean code for falsey or undefined values
- not adding on event listener when the value is undefined

Personally, I care more about the second.

**Comment by jelhan** [comment]
> > An alternative solution might be returning a noop function if `{{fn}}` helper is called without any argument.
> 
> This would pose the exact same hazard, just in a different place.

Are you sure? I suppose that `{{fn}}` returns a noop function if called without any argument. Calling it with an argument which is not a function (e.g. `{{fn undefined}}`), should still be invalid and throw.

**Comment by NullVoxPopuli** [comment]
fn already solely means 'partial application' though?

**Comment by jelhan** [comment]
> fn already solely means 'partial application' though?

You are right that the meaning of `{{fn}}` helper would slightly change. I propose that it could be summarized like this:

> The `{{fn}}` helper returns a new function.
>
> You may pass a callable as first positional argument. The returned function calls that callable with all additional positional arguments given.
>
> If no positional argument is given a noop function is retuned.

**Comment by buschtoens** [comment]
@jelhan The `{{fn}}` helper expects the first argument to always be a function:

https://github.com/emberjs/ember.js/blob/b049c37e66587718cef38cde63b59b0789505d43/packages/%40ember/-internals/glimmer/lib/helpers/fn.ts#L82-L92
https://github.com/emberjs/ember.js/blob/b049c37e66587718cef38cde63b59b0789505d43/packages/%40ember/-internals/glimmer/lib/helpers/fn.ts#L97-L102

Allowing `{{fn undefined}}` to create a no-op function, like `{{optional}}` does, IMO is dangerous, as it opens the door for typo bugs, much in the same way that it would for `{{on}}`.

Going back to the two design goals by @NullVoxPopuli:

> - clean code for falsey or undefined values
> - not adding on event listener when the value is undefined

The first one IMO is sufficiently covered by `{{optional}}`. The second one can be solved either, when RFC #432 is implemented, as explained above, or by adding some param like `optional=true` to the `{{on}}` modifier:

```hbs
<button {{on "click" @onClick optional=true}}>
```

I agree very much, that both goals are super desirable, but I don't think increasing the API surface area or loosening the safety net is a good idea.

**Comment by wagenet** [comment]
I'm closing this due to inactivity. This doesn't mean that the idea presented here is invalid, but that, unfortunately, nobody has taken the effort to spearhead it and bring it to completion. Please feel free to advocate for it if you believe that this is still worth pursuing. Thanks!

**Comment by NullVoxPopuli** [comment]
this problem is now solved, since modifiers can be conditional:

[See this demo](https://limber.glimdown.com/?t=%60%60%60gjs+live+preview%0Aimport+%7B+on+%7D+from+%27%40ember%2Fmodifier%27%3B%0A%0Afunction+sayHello%28%29+%7B%0A++alert%28%27hello%27%29%3B%0A%7D%0A%0A%3Ctemplate%3E%0A++%3Cbutton+%7B%7B%28if+true+%28modifier+on+%22click%22+sayHello%29%29%7D%7D%3E%0A++++yes%0A++%3C%2Fbutton%3E%0A%0A++%3Cbutton+%7B%7B%28if+false+%28modifier+on+%22click%22+sayHello%29%29%7D%7D%3E%0A++++no%0A++%3C%2Fbutton%3E%0A%3C%2Ftemplate%3E%0A%60%60%60%0A)

```js
import { on } from '@ember/modifier';

function sayHello() {
  alert('hello');
}

<template>
  <button {{(if true (modifier on "click" sayHello))}}>
    yes
  </button>

  <button {{(if false (modifier on "click" sayHello))}}>
    no
  </button>
</template>
```

**Comment by NullVoxPopuli** [comment]
Today, we can achieve the same effect via conditional modifier, [demo here](https://limber.glimdown.com/?t=%60%60%60gjs+live+preview%0Aimport+%7B+on+%7D+from+%27%40ember%2Fmodifier%27%3B%0A%0Afunction+sayHello%28%29+%7B%0A++alert%28%27hello%27%29%3B%0A%7D%0A%0A%3Ctemplate%3E%0A++%3Cbutton+%7B%7B%28if+true+%28modifier+on+%22click%22+sayHello%29%29%7D%7D%3E%0A++++yes%0A++%3C%2Fbutton%3E%0A%0A++%3Cbutton+%7B%7B%28if+false+%28modifier+on+%22click%22+sayHello%29%29%7D%7D%3E%0A++++no%0A++%3C%2Fbutton%3E%0A%3C%2Ftemplate%3E%0A%60%60%60%0A)

```js
import { on } from '@ember/modifier';

function sayHello() {
  alert('hello');
}

<template>
  <button {{(if true (modifier on "click" sayHello))}}>
    yes
  </button>

  <button {{(if false (modifier on "click" sayHello))}}>
    no
  </button>
</template>
```
